src/clarity_vm/clarity.rs

```diff
@@ -275,27 +281,38 @@ impl ClarityInstance {
     }
 
     /// Returns the Stacks epoch of the burn block that elected `stacks_block`
+    ///
+    /// 1) If this is a test, and the burn_height is set to 0, return Epoch20. (Issue #2907)
+    /// 2) Otherwise return the epoch found according to the height.
+    /// 3) Default to Epoch20, if no epoch found in (2).
     fn get_epoch_of(
         stacks_block: &StacksBlockId,
         header_db: &dyn HeadersDB,
         burn_state_db: &dyn BurnStateDB,
-    ) -> StacksEpochId {
-        match header_db.get_burn_block_height_for_block(stacks_block) {
+    ) -> StacksEpoch {
+        // Step 1: Try to find the epoch according to block.
+        let epoch_found = match header_db.get_burn_block_height_for_block(stacks_block) {
             Some(burn_height) => {
-                // special case the Stacks 2.0 genesis block -- it occurs at a zero burn block height
-                if burn_height == 0 {
-                    GENESIS_EPOCH
+                // We hard-code this special case to use Epoch20 for `burn_height` 0 to keep unit
+                // tests passing. Issue #2907 is open to remove this case.
+                if cfg!(test) && burn_height == 0 {
+                    None
                 } else {
-                    burn_state_db
-                        .get_stacks_epoch(burn_height)
-                        .expect(&format!(
-                            "Failed to get Stacks epoch for height = {}",
-                            burn_height
-                        ))
-                        .epoch_id
+                    Some(burn_state_db.get_stacks_epoch(burn_height).expect(&format!(
+                        "Failed to get Stacks epoch for height = {}",
+                        burn_height
+                    )))
                 }
             }
-            None => GENESIS_EPOCH,
+            None => None,
+        };
+
+        // Step 2: If not found, default to Epoch20.
+        match epoch_found {
+            Some(epoch) => epoch,
+            None => burn_state_db
+                .get_stacks_epoch_by_epoch_id(&GENESIS_EPOCH)
+                .expect("Failed to get Stacks epoch for GENESIS_EPOCH"),
```

Comment:
Continuing the thread from the comment chain on the commit with @jcnelson and @gregorycoppola.

Right now, this performs the special case handling only in `#cfg(test)`. This special case handling was actually already present in my previous PR, so I think it's somewhat orthogonal to the work of this PR. I opened an issue #2907 for resolving this. I do not think that this issue should block this PR, though.

Does that sound okay to you, @jcnelson ?

Reply:
> This special case handling was actually already present in my previous PR

I would have raised this issue then if I had caught it then.

I'm very uncomfortable as to why we can't answer the question "why does `burn_state_db.get_stacks_epoch(0)` return `None`?"  Aren't epochs defined for each and every block height in the `u64` domain?

Reply:
These codepaths only occur when using mocked burn state databases. Fixing this will involve synchronizing all of the "first burnchain height" parameters passed to different unit tests with their instantiated mock header dbs. It's possible, but it's a task that's pretty unrelated to this PR.

Reply:
Okay, but I'd like to ship #2907 in `next-costs` if at all possible.  Adding test-specific special cases to production code paths -- especially consensus-critical code paths like this -- makes me deeply uncomfortable.

Reply:
Yeah, it should be possible -- I already started working on the patchset for the change. I think the best approach is to do something like:

1. Create a new NullHeadersDB and NullBurnstateDB which are like UnitTestHeadersDB and UnitTestBurnstateDB, which are `#[cfg(test)]`
2. Make NullHeadersDB and NullBurnstateDB actually just always panic on invocation. I need to enumerate all of their usage outside of testing, but I'm pretty sure its just in the evaluation of cost functions (which panic on database lookups: cost functions should never invoke db methods).
3. Make sure that unit test headers/burn dbs return correctly enough mocked information to pass tests.

Reply:
Okay b90a5ba20 contains the changes:

1. Fix the panic while panicking issue which made running down test failures of this kind really annoying.
2. Explicit special case for when the parent is the "FIRST_STACKS_BLOCK". That "block" _isn't ever_ elected, so it would fail to lookup a burn block height (or the height wouldn't be sane). I don't see a way around this special case.
3. Add `UnitTest..` versions of the mock burn and headers interfaces. These versions return some sane-ish mock results for block height.
4. Alter the read-only connections to use the provided `at_block`'s _evaluated_ epoch version, rather than trying to figure out the epoch that elected the block. The difference is subtle, but I believe this is the desired behavior (the read-only connections will be evaluated under the same rules that `at_block` was evaluated under, rather than the rules that `at_block`'s descendants will be evaluated under).

Reply:
Re (4), does this go against what we were thinking for 2.1 -- i.e. the `at-block` code body runs with the VM rules of the contract containing the `at-block` invocation, instead of the VM rules of the block indicated by `at-block`?

Reply:
Re (4), so I understand correctly (since this is what the code is now doing) -- the VM rules that get applied by an `at-block` are the rules in which the `at-block` is running, not the rules that were in place in the ancestor block in which the `at-block` code body is meant to be run?  I.e. if `(at-block X (print "hello"))` runs in block Y, then the rules for Y, not X, are applied to `(print "hello")`, right?

Reply:
Ah, point (4) above doesn't impact `at-block`, rather I'm talking about the `eval_read_only` style methods that things like RPCs use (and the chains coordinator uses to get the amount of liquid stx). Those methods open a read-only connection, but need to supply an EpochId for the GlobalContext.

When it comes to 2.1, we'll need to be careful to make sure that it uses the same method of figuring the epoch that 2.05 uses -- basically, the epoch of the parent governs the current epoch.

Reply:
The one thing left I want to highlight is the change in 85a42f0816857d8c307acf3aa8cf215711d29680. The goal there was just to eliminate an uneccessary usage of the NULL_BURN_STATE_DB by doing the `rollback_unconfirmed` invocation directly. Does that change make sense @jcnelson ?

Reply:
Yes, I think that's fine.  As long as the call to `datastore.rollback_unconfirmed()` is preceded by a `datastore.begin_unconfirmed()`, then the behavior should be the same -- i.e. the inner MARF will be pointed to the unconfirmed state trie, so when we roll it back, the unconfirmed state trie gets deleted.